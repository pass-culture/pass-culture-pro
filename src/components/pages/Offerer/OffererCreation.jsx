import PropTypes from 'prop-types'
import React, { PureComponent } from 'react'

import { OffererClass } from './OffererClass'
import HeroSection from '../../layout/HeroSection/HeroSection'
import Main from '../../layout/Main'
import { Form } from 'react-final-form'
import Siren from './Siren/Siren'
import Name from './Name'
import Address from './Address'
import createDecorator from "final-form-calculate"
import { bindAddressAndDesignationFromSiren } from "./decorators/bindSirenFieldToDesignation"
import PostalCode from "./PostalCode"
import City from "./City"

class OffererCreation extends PureComponent {


  componentDidMount() {
    const { getOfferer } = this.props
    getOfferer()
  }

  handleSubmit = (values) => {
    const { createNewOfferer } = this.props
    createNewOfferer(values, this.onHandleFail)
  }

  onHandleDataRequest = (handleSuccess, handleFail) => {
    const { getOfferer, getUserOfferers, query } = this.props
    const { isCreatedEntity } = query.context()

    if (!isCreatedEntity) {
      getOfferer(handleFail, handleSuccess)
      getUserOfferers()
      return
    }
    handleSuccess()
  }

  onHandleFail = () => {
    const { showNotification } = this.props
    showNotification('Vous étes déjà rattaché à cette structure.', 'danger')
  }

  onRender = ({ handleSubmit, submitting, pristine }) => {
    return (
      <form onSubmit={handleSubmit}>
        <div className="section">
          <div className="field-group">

            <Siren />
            <Name />
            <Address />
            <PostalCode />
            <City />

            <button
              className="button is-primary is-medium"
              disabled={submitting || pristine}
              type="submit"
            >
              {'Valider'}
            </button>
          </div>
        </div>
      </form>
    )
  }

  createDecorators = () => {
    const addressAndDesignationFromSirenDecorator = createDecorator({
      field: 'siren',
      updates: bindAddressAndDesignationFromSiren,
    })

    return [ addressAndDesignationFromSirenDecorator ]
  }

  render() {
    const { offerer} = this.props

    return (
      <Main
        backTo={{ label: 'Vos structures juridiques', path: '/structures' }}
        name="offerer"
      >
        <HeroSection
          subtitle={offerer.name}
          title="Structure"
        >
          <p className="subtitle">
            {'Détails de la structure rattachée, des lieux et des fournisseurs de ses offres.'}
          </p>
        </HeroSection>

        <Form
          decorators={this.createDecorators()}
          initialValues={offerer}
          onSubmit={this.handleSubmit}
          render={this.onRender}
        />
      </Main>
    )
  }
}

OffererCreation.propTypes = {
  createNewOfferer: PropTypes.func.isRequired,
  getOfferer: PropTypes.func.isRequired,
  getUserOfferers: PropTypes.func.isRequired,
  offerer: PropTypes.instanceOf(OffererClass).isRequired,
  query: PropTypes.shape().isRequired,
  showNotification: PropTypes.func.isRequired,
}

export default OffererCreation
