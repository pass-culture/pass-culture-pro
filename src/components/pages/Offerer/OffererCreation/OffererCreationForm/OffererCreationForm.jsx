import React from 'react'
import PropTypes from 'prop-types'
import Siren from '../Fields/Siren/Siren'
import Name from '../Fields/Name'
import Address from '../Fields/Address'
import PostalCode from '../Fields/PostalCode'
import City from '../Fields/City'
import { NavLink } from "react-router-dom"

function OffererCreationForm({ handleSubmit, invalid, pristine }) {
  return (
    <form onSubmit={handleSubmit}>
      <div className="section">
        <div className="field-group">
          <Siren />
          <Name />
          <Address />
          <PostalCode />
          <City />
        </div>
        <div
          className="offerer-form-validation"
          style={{ justifyContent: 'space-between' }}
        >
          <div className="control">
            <NavLink
              className="button is-secondary is-medium"
              to="/structures"
            >
              {'Retour'}
            </NavLink>
          </div>
          <div className="control">
            <button
              className="button is-primary is-medium"
              disabled={invalid || pristine}
              type="submit"
            >
              {'Valider'}
            </button>
          </div>
        </div>
      </div>
    </form>
  )
}

OffererCreationForm.propTypes = {
  handleSubmit: PropTypes.func.isRequired,
  invalid: PropTypes.bool.isRequired,
  pristine: PropTypes.bool.isRequired
}

export default OffererCreationForm
