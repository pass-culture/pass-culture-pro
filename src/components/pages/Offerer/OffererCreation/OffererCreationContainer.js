import { removeWhitespaces } from 'react-final-form-utils'
import { connect } from 'react-redux'
import { compose } from 'redux'
import { requestData } from 'redux-saga-data'

import withTracking from 'components/hocs/withTracking'
import { showNotification } from 'store/reducers/notificationReducer'
import { selectIsFeatureActive } from 'store/selectors/data/featuresSelectors'
import { selectCurrentUser } from 'store/selectors/data/usersSelectors'

import OffererCreation from './OffererCreation'

export function mapStateToProps(state) {
  return {
    currentUser: selectCurrentUser(state),
    isNewHomepageActive: selectIsFeatureActive(state, 'PRO_HOMEPAGE'),
  }
}

export const mapDispatchToProps = dispatch => ({
  createNewOfferer: (payload, onHandleFail, onHandleSuccess) => {
    const { siren } = payload
    dispatch(
      requestData({
        apiPath: `/offerers`,
        method: 'POST',
        body: { ...payload, siren: removeWhitespaces(siren) },
        handleFail: onHandleFail,
        handleSuccess: onHandleSuccess,
      })
    )
  },
  showNotification: (message, type) => {
    dispatch(
      showNotification({
        text: message,
        type: type,
      })
    )
  },
})

export const mergeProps = (stateProps, dispatchProps, ownProps) => {
  return {
    ...stateProps,
    ...dispatchProps,
    ...ownProps,
    trackCreateOfferer: createdOffererId => {
      ownProps.tracking.trackEvent({ action: 'createOfferer', name: createdOffererId })
    },
    redirectAfterSubmit: createdOffererId => {
      ownProps.history.replace(
        stateProps.isNewHomepageActive ? `/accueil?structure=${createdOffererId}` : '/structures'
      )
    },
  }
}

export default compose(
  withTracking('Offerer'),
  connect(mapStateToProps, mapDispatchToProps, mergeProps)
)(OffererCreation)
