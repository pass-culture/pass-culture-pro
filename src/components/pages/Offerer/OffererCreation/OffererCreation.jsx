import createDecorator from 'final-form-calculate'
import PropTypes from 'prop-types'
import React, { PureComponent } from 'react'
import { Form } from 'react-final-form'

import AppLayout from 'app/AppLayout'
import PageTitle from 'components/layout/PageTitle/PageTitle'
import Titles from 'components/layout/Titles/Titles'

import { bindAddressAndDesignationFromSiren } from './decorators/bindSirenFieldToDesignation'
import OffererCreationForm from './OffererCreationForm/OffererCreationForm'

class OffererCreation extends PureComponent {
  handleSubmit = values => {
    const { createNewOfferer } = this.props
    createNewOfferer(values, this.onHandleFail, this.onHandleSuccess)
  }

  onHandleSuccess = (_, action) => {
    const { trackCreateOfferer, redirectAfterSubmit } = this.props
    const { payload } = action
    const createdOffererId = payload.datum.id

    trackCreateOfferer(createdOffererId)
    redirectAfterSubmit()
  }

  onHandleFail = () => {
    const { showNotification } = this.props
    showNotification('Vous étes déjà rattaché à cette structure.', 'error')
  }

  createDecorators = () => {
    const addressAndDesignationFromSirenDecorator = createDecorator({
      field: 'siren',
      updates: bindAddressAndDesignationFromSiren,
    })

    return [addressAndDesignationFromSirenDecorator]
  }

  render() {
    const { isNewHomepageActive } = this.props
    return (
      <AppLayout
        layoutConfig={{
          backTo: isNewHomepageActive
            ? { label: 'Accueil', path: '/accueil' }
            : { label: 'Structures juridiques', path: '/structures' },
          pageName: 'offerer',
        }}
      >
        <PageTitle title="Créer une structure" />
        <Titles title="Structure" />

        <Form
          backTo={isNewHomepageActive ? '/accueil' : '/structures'}
          component={OffererCreationForm}
          decorators={this.createDecorators()}
          onSubmit={this.handleSubmit}
        />
      </AppLayout>
    )
  }
}

OffererCreation.propTypes = {
  createNewOfferer: PropTypes.func.isRequired,
  isNewHomepageActive: PropTypes.bool.isRequired,
  redirectAfterSubmit: PropTypes.func.isRequired,
  showNotification: PropTypes.func.isRequired,
  trackCreateOfferer: PropTypes.func.isRequired,
}

export default OffererCreation
