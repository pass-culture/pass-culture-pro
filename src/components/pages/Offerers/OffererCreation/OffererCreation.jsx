import createDecorator from 'final-form-calculate'
import PropTypes from 'prop-types'
import React, { PureComponent } from 'react'
import { Form } from 'react-final-form'

import AppLayout from 'app/AppLayout'
import PageTitle from 'components/layout/PageTitle/PageTitle'
import Titles from 'components/layout/Titles/Titles'
import { bindAddressAndDesignationFromSiren } from 'repository/siren/bindSirenFieldToDesignation'

import OffererCreationForm from './OffererCreationForm/OffererCreationForm'

class OffererCreation extends PureComponent {
  handleSubmit = values => {
    const { createNewOfferer } = this.props
    createNewOfferer(values, this.onHandleFail, this.onHandleSuccess)
  }

  onHandleSuccess = (_, action) => {
    const { trackCreateOfferer, redirectAfterSubmit } = this.props
    const { payload } = action
    const createdOffererId = payload.datum.id

    trackCreateOfferer(createdOffererId)
    redirectAfterSubmit(createdOffererId)
  }

  onHandleFail = () => {
    const { showNotification } = this.props
    showNotification('Vous étes déjà rattaché à cette structure.', 'error')
  }

  createDecorators = () => {
    const addressAndDesignationFromSirenDecorator = createDecorator({
      field: 'siren',
      updates: bindAddressAndDesignationFromSiren,
    })

    return [addressAndDesignationFromSirenDecorator]
  }

  render() {
    return (
      <AppLayout
        layoutConfig={{
          backTo: { label: 'Accueil', path: '/accueil' },
          pageName: 'offerer',
        }}
      >
        <PageTitle title="Créer une structure" />
        <Titles title="Structure" />

        <Form
          backTo="/accueil"
          component={OffererCreationForm}
          decorators={this.createDecorators()}
          onSubmit={this.handleSubmit}
        />
      </AppLayout>
    )
  }
}

OffererCreation.propTypes = {
  createNewOfferer: PropTypes.func.isRequired,
  redirectAfterSubmit: PropTypes.func.isRequired,
  showNotification: PropTypes.func.isRequired,
  trackCreateOfferer: PropTypes.func.isRequired,
}

export default OffererCreation
