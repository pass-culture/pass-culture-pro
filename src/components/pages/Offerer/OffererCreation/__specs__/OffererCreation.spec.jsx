import OffererCreation from '../OffererCreation'
import React from 'react'
import {shallow} from 'enzyme'
import {Form} from 'react-final-form'
import OffererCreationForm from '../OffererCreationForm/OffererCreationForm'

describe('src | components | pages | Offerer | OffererCreation | OffererCreation', () => {
  let props

  beforeEach(() => {
    props = {
      createNewOfferer: jest.fn(),
      showNotification: jest.fn(),
      closeNotification: jest.fn(),
      trackCreateOfferer: jest.fn(),
      redirectToOfferersList: jest.fn()
    }
  })

  it('should match snapshot', () => {
    // when
    const wrapper = shallow(<OffererCreation {...props} />)

    // then
    expect(wrapper).toMatchSnapshot()
  })

  describe('event tracking', () => {
    it('should track offerer creation', () => {
      // given
      const state = {}
      const action = {
        payload: {
          datum: {
            id: 'Ty5645dgfd',
          },
        },
      }
      const wrapper = shallow(<OffererCreation {...props} />)

      // when
      wrapper.instance().onHandleSuccess(state, action)

      // then
      expect(props.trackCreateOfferer).toHaveBeenCalledWith('Ty5645dgfd')
    })
  })

  describe('render', () => {
    it('should render a OffererCreation component with default props', () => {
      // when
      const wrapper = shallow(<OffererCreation {...props} />)

      // then
      expect(wrapper.prop('createNewOfferer')).toBe()
      expect(wrapper.prop('showNotification')).toBe()
    })

    it('should display "Structure" title', () => {
      // given
      const wrapper = shallow(<OffererCreation {...props} />)

      // when
      const heroSection = wrapper.find('HeroSection').props()

      // then
      expect(heroSection.title).toStrictEqual('Structure')
    })

    it('should display offerer creation form', () => {
      // when
      const wrapper = shallow(<OffererCreation {...props} />)

      // then
      const form = wrapper.find(Form)
      const renderProp = form.prop('render')
      expect(renderProp).toBe(OffererCreationForm)
    })
  })

  describe('componentWillUnmount', () => {
    it('should close notification', () => {
      // when
      const wrapper = shallow(<OffererCreation {...props} />)
      wrapper.unmount()

      // expect
      expect(props.closeNotification).toHaveBeenCalledWith()
    })
  })
})
