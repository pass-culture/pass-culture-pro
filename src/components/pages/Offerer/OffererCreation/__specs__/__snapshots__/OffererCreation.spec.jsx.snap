// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`src | components | pages | Offerer | OffererCreation | OffererCreation should match snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <OffererCreation
    closeNotification={[MockFunction]}
    createNewOfferer={[MockFunction]}
    redirectToOfferersList={[MockFunction]}
    showNotification={[MockFunction]}
    trackCreateOfferer={[MockFunction]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "backTo": Object {
        "label": "Vos structures juridiques",
        "path": "/structures",
      },
      "children": Array [
        <Titles
          action={null}
          subtitle={null}
          title="Structure"
        />,
        <ReactFinalForm
          decorators={
            Array [
              [Function],
            ]
          }
          onSubmit={[Function]}
          render={[Function]}
        />,
      ],
      "name": "offerer",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "action": null,
          "subtitle": null,
          "title": "Structure",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "decorators": Array [
            [Function],
          ],
          "onSubmit": [Function],
          "render": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "backTo": Object {
          "label": "Vos structures juridiques",
          "path": "/structures",
        },
        "children": Array [
          <Titles
            action={null}
            subtitle={null}
            title="Structure"
          />,
          <ReactFinalForm
            decorators={
              Array [
                [Function],
              ]
            }
            onSubmit={[Function]}
            render={[Function]}
          />,
        ],
        "name": "offerer",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "action": null,
            "subtitle": null,
            "title": "Structure",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "decorators": Array [
              [Function],
            ],
            "onSubmit": [Function],
            "render": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    Symbol(enzyme.__providerValues__): undefined,
  },
  Symbol(enzyme.__providerValues__): Map {},
  Symbol(enzyme.__childContext__): null,
}
`;
